### ISSUE TITLE
How to look up old articles?

### ISSUE THREAD

**Issue by jschlatow** [genodelabs/genodians.org#34]
The index page currently lists the 100 most recent articles. This makes looking up older articles (e.g. the Goa Unix tutorial) or even searching for certain keywords rather complicated.

I believe adding an archive page to the blog where all articles are listed would be the easiest solution. Any opinions on this?

**Comment by nfeske** [comment]
The reason behind the limit was that the extraction of the title and the first paragraph for each article is not cheap. So I capped it. That said, I have no objection against an archive page - if you like to give it a try.

BTW, the limit does not exist on the authors page. If you know the author of an article, you should be able to find it there. Another life hack was mentioned in cnuke's latest posting, using `site:genodians.org` in a regular search engine. But these are excuses. If you want to improve the situation, that's very welcome!

**Comment by jschlatow** [comment]
Commit a8d545c implements this idea.

Since you mentioned extracting the summaries for each article is not cheap, I also tried to optimise the Makefile a little so that each summary is only generated once. See a6f2c64.

**Comment by jschlatow** [comment]
Unfortunately, during test-running, I noticed that the index generation takes very long despite the aforementioned optimisation. I had a more detailed look and noticed that merely executing `echo`s and `cat`s in a make recipe already eats up a lot of processing time. Make spawns a new bash for each of those and spawning a bash is pretty slow. I tried to narrow this down a bit further by taking my unix_shell scenario and doing the following in a bash script:

```
for i in {1..5}; do
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
	bash -c 'echo "one" > /dev/null'
done
```

Executing this script takes more than 20 seconds (on base-linux). The same holds when I replace the sub-bash with `cat`, hence I think this is unrelated to the size of the binary. 

Long story short: As a workaround, I tried to reduce the number of SHELL invocations in the Makefile. See bdbeece.

**Comment by jschlatow** [comment]
I took a few more measurements since the results I got from the aforementioned bash script still puzzle me.

- Running the same script in Sculpt's inspect view is pretty fast (1s).
- In my Linux VM on Sculpt it takes ~20s on base-linux (using run/bash on the current master branch)
- If I run my Linux natively, the effect vanishes.

Since I suspected `fork()` to be slow, I wrote a simple test. Irrespective of how I start up my Linux, however, forking is reasonably fast on base-linux. Yet, it is terribly slow on base-nova (running in qemu).

In summary, it doesn't look as bad as I thought yesterday.

**Comment by nfeske** [comment]
Thanks for taking the initiative for the investigation.

One thing to keep in mind is that the forking of a non-trivial process (like bash) benefits a lot from cached_fs_rom. (on base-linux, cached_fs_rom is always replaced by fs_rom because core's RM service is not usable on base-linux) Hence, usually, scenarios that depend on fork run much slower on base-linux than on base-nova. I suspect that your observation of the opposite comes down to the use of qemu for running NOVA?

From a quick glance, genodians.org uses cached_fs_rom for the generate step.
